File: Python/stacks_and_queues/stack_min2.py

from stack import Stack

class Node():
    def __init__(self, value):
        self.minimum = None
        self.value = value

class ExtStack(Stack):
    def __init__(self):
        super().__init__()

    def push(self, value):
        if self.is_empty():
            super().push(Node(value))
            self.items[-1].minimum = self.items[-1]
        elif value <= self.items[-1].minimum.value:
            super().push(Node(value))
            self.items[-1].minimum = self.items[-1]
        else:
            super().push(Node(value))
            self.items[-1].minimum = self.items[-2].minimum

    def min(self):
        if self.is_empty():
            print("Stack is empty.")
            exit(1)
        return self.items[-1].minimum.value


if __name__ == "__main__":
    stack = ExtStack()
    for v in [5, 10, 4, 9, 3, 3, 8, 2, 2, 7, 6]:
        stack.push(v)
        print(f"Pushed value: {v}, min value: {stack.min()}")
    while not stack.is_empty():
        v = stack.pop()
        if stack.is_empty():
            print(f"Popped value: {v}, stack is empty")
        else:
            print(f"Popped value: {v}, min value: {stack.min()}")

File: C++/stacks_and_queues/stack_min2.cpp
#include "stack.h"

template<typename T>
class ExtStack : public Stack<T> {
public:
    struct Node : public Stack<T>::Node {
        Node* minimum;
        Node *next;
        Node(const T& v, Node* n = nullptr) : Stack<T>::Node(v, n), minimum(nullptr) {}
        Node(T&& v, Node* n = nullptr) : Stack<T>::Node(std::move(v), n), minimum(nullptr) {}
    };

    Node *top;    
    template <typename U>
    void push(U &&value);
    int min() const;
};

template<typename T>
template <typename U>
void ExtStack<T>::push(U &&value) 
{
    if(this->isEmpty())
    {
        this->top = new Node(std::forward<U>(value));
        this->top->minimum = this->top;
    }
    else if(value <= this->top->minimum->value)
    {
        this->top = new Node(std::forward<U>(value), this->top);
        this->top->minimum = this->top;
    }
    else
    {
        this->top = new Node(std::forward<U>(value), this->top);
        this->top->minimum = this->top->next->minimum;
    }

    this->stackSize++;
}

template<typename T>
int ExtStack<T>::min() const
{
    if(this->isEmpty())
    {
        std::cout << "Stack is empty." << std::endl;
        exit(1);
    }
    return (this->top->minimum->value);
}

int main()
{
    ExtStack<int> stack;
    for (auto v : {5, 10, 4, 9, 3, 3, 8, 2, 2, 7, 6})
    {
        stack.push(v);
        std::cout << "Pushed value: " << v << ", min value: " << stack.min() << std::endl;
    }
    while (!stack.isEmpty())
    {
        auto v = stack.pop();
        std::cout << "Popped value: " << v;
        if (stack.isEmpty())
            std::cout << ", stack is empty" << std::endl;
        else
            std::cout << ", min value: " << stack.min() << std::endl;
    }
    return 0;
}
